lib/state_machine/state_machine.hpp

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include "consts.hpp"
+#include "state.hpp"
+#include "types.hpp"
+
+#include <array>
+#include <optional>
+
+namespace hyped::state_machine {
+
+class StateMachine {
+ public:
+  StateMachine() { current_state = State::kIdle; };
+
+  void transition(const Message message);
+
+  std::optional<Message> checkTransition();
+
+  void reset();
+
+  // TODO change transitions to actual transitions
+  std::array<Transition, kNumTransitions> transitions{
```

Comment:
I don't agree with the data structure here.

Think about "What do I want to do?". I claim there is exactly one way in which you are going to look things up. The current lookup time is `O(n * m)` where `n` is the number of states and `m` the number of messages; this should be constant. Talk to my IADS tutees, they should know how to make this better.

Reply:
Would an unordered map be more appropriate for this?

Reply:
certainly

you probably want to use a key like `State * Message` but put that in a struct like `Event` and then figure out how to hash it.

Reply:
For the hashing, i think we can just copy over Boost's hash_combine() and put it in utils::hash.

Reply:
I mean we could also just actually depend on boost, like approximately any other C++ project :P

Reply:
But there's another thing: you either need to make a custom hasher and carry it around or invoke some black magic which I unfortunately forgot (it involved defining an empty template inside the `std` namespace).

Ofc you might say "but why not use `std::pair<S,T>` and the answer is because the actual code (outside the types) will be ugly if we do this.

Reply:
that's the magic i was talking about